src/lem/circuit.rs

```diff
@@ -166,117 +171,121 @@ fn allocate_img_for_slot<F: LurkField, CS: ConstraintSystem<F>>(
+                        cs.namespace(|| format!("component {component_idx} slot {slot}")),
                         || z_ptr.tag_field(),
                     ));
-                    // allocate third component
+
+                    component_idx += 1;
+
+                    // allocate pointer hash
                     preallocated_preimg.push(AllocatedNum::alloc_infallible(
-                        cs.namespace(|| format!("component 2 slot {slot}")),
+                        cs.namespace(|| format!("component {component_idx} slot {slot}")),
                         || *z_ptr.value(),
                     ));
+
+                    component_idx += 1;
                 }
-                SlotData::F(a) => {
-                    preallocated_preimg.push(AllocatedNum::alloc_infallible(
-                        cs.namespace(|| format!("component 0 slot {slot}")),
-                        || *a,
-                    ));
-                }
             }
+            SlotData::FPtr(f, ptr) => {
+                let z_ptr = store.hash_ptr(ptr)?;
+                // allocate first component
+                preallocated_preimg.push(AllocatedNum::alloc_infallible(
+                    cs.namespace(|| format!("component 0 slot {slot}")),
+                    || *f,
+                ));
+                // allocate second component
+                preallocated_preimg.push(AllocatedNum::alloc_infallible(
+                    cs.namespace(|| format!("component 1 slot {slot}")),
+                    || z_ptr.tag_field(),
+                ));
+                // allocate third component
+                preallocated_preimg.push(AllocatedNum::alloc_infallible(
+                    cs.namespace(|| format!("component 2 slot {slot}")),
+                    || *z_ptr.value(),
+                ));
+            }
+            SlotData::F(a) => {
+                preallocated_preimg.push(AllocatedNum::alloc_infallible(
+                    cs.namespace(|| format!("component 0 slot {slot}")),
+                    || *a,
+                ));
+            }
+        }
 
-            // Allocate the image by calling the arithmetic function according
-            // to the slot type
-            let preallocated_img =
-                allocate_img_for_slot(cs, &slot, preallocated_preimg.clone(), store)?;
+        // Allocate the image by calling the arithmetic function according
+        // to the slot type
+        let preallocated_img =
+            allocate_img_for_slot(cs, &slot, preallocated_preimg.clone(), store)?;
+
+        Ok((preallocated_preimg, preallocated_img))
+    } else {
+        let preallocated_preimg: Vec<_> = (0..slot_type.preimg_size())
+            .map(|component_idx| {
+                AllocatedNum::alloc_infallible(
+                    cs.namespace(|| format!("component {component_idx} slot {slot}")),
+                    || F::ZERO,
+                )
+            })
+            .collect();
```

Comment:
Although I don't think it's introduced in this PR, it seems that the handling of preimage values is smudging the bellpepper API. When we are synthesizing shapes without data, we should not be specifying concrete values. Even if we get away with it here, this is sloppy and may end up biting us.

Reply:
My point is not that the above code will *always* run in a context where no real value exist — only that I think it sometimes will.

Reply:
Maybe I'm missing the point, but this is not to synthesize the shape without data. It's to synthesize slots that weren't visited during interpretation with dummy data

Reply:
Right, which is what I mentioned in my second comment. But the point is that when synthesizing without data, one of the two branches of this conditional will have to be taken, and they both allocate infallibly, as far as I can tell. Is there some third code path that is used to synthesize when creating a shape? If so, that's also not the standard usage.

Reply:
No, this is the only code that's used to create the shape. But I think this is getting into territory @gabriel-barrett understands better

Reply:
It has come up elsewhere, and I'm reiterating the point here.

Reply:
Can you create an issue for it please? I don't think I understand the problem well enough yet.